### ISSUE TITLE
"Equalizing/Normalizing" stacks across characters (testing/GUI help?)

### ISSUE THREAD

**Issue by Mulchman** [dasilva333/TowerGhostForDestiny#122]
I don't know where I saw the feature request - on TowerGhostForDestiny/DIV/DIM (I 'watch' all but there's lots of noise on some days so I mass delete every once in a while) - but someone at some point wanted the ability to automatically equalize/normalize materials/consumables across their characters.

I made a fork (https://github.com/Mulchman/TowerGhostForDestiny) and wrote some code to do this but I don't think it's at the pull request stage yet. (Or is this even a feature TGfD would be up for including?) I'm looking for a few people to test/review the code/functionality and offer any input. This was my first time writing JavaScript code so I'm completely ignorant of best practices and what not. Feel free to be brutally honest; you won't hurt my feelings.

As far as using it - there's a toggle under settings - "Equalize Stacks" - that when enabled and you click on a consumable or material, will start moving things around to distribute whatever you clicked on among your other characters. The vault is excluded.  (Should it be included or an option?) I'm also thinking a GUI would be nice before it starts blunt force changing your inventory but I'm awful at design and my HTML level is rudimentary at best.  And like with transferring any consumable/material in TGfD, you have to refresh afterwards so stacks stack up correctly and internal amounts and what not become correct again.

Anyway, to recap, the things I was looking for input on-

* code quality? Am I doing dumb things?
* where should the option for this be? On the move pop-up? A toggleable setting like now? A special 'stack management' menu? (I have an idea of another feature to add that's stack management related as well - a 'move all of type' to a player/vault.)
* should the vault be included when equalizing/normalizing? like, should the vault get a cut of the distribution? or should the vault be included in the sense of the items in the vault should be emptied and distributed among characters?
* would some kind of GUI be nice before the blunt force operations happen?

**Comment by dasilva333** [comment]
Cool man good job, submit a pull request now even if it's not ready so I can review/compare and provide you some tips. I don't think there should be a GUI but I do think there should be a confirm before it happens so the flow I think should be:

Settings > Normalize Stacks > Are you sure you want to balance out the following materials evenly across all characters?

[x] Heavy Ammo Synth
[x] Special Ammo Synth
[ ] Primary Synth (notice how I unchecked one?)

[Confirm][Cancel]

Also regarding the inner works moving/merging stacks is known to be buggy (as you probably saw in the code) so this feature might have to wait until the core bugs are ironed out, basically lets suppose you do the following:
1. merge a stack
2. move the stack
3. split the stack
4. try to merge it again (might not work, or any variation or merge, split, merge)

**Comment by Mulchman** [comment]
I think I did the pull request properly. I definitely like your idea of it being a button you press once rather than a mode (ie. destinyDbMode). For the confirm dialog, I really suck at HTML, and could definitely use some help there.

**Comment by dasilva333** [comment]
@Mulchman looked over the code everything looks legit, I also like better how you did it than my original idea, it can get messy trying to normalize everything at once, your method allows for methodical balancing which I like better. As for the UI I think you did fine by providing the user with prompts in case something goes wrong, I would just add one bootstrapdialog.alert from the get go that shows you what the end result is going to be for that one item, confirm and let it do it's thing.

**Comment by Mulchman** [comment]
Moving the conversation to the pull request. https://github.com/dasilva333/TowerGhostForDestiny/pull/123

**Comment by lanejcf** [comment]
This is working well and a great addition - Would it be possible to add a normalize icon to the move popup that appears when you select a consumable/material to save having to go to the menu and toggle the mode ?

**Comment by Mulchman** [comment]
I do agree the menu option makes things a little obtuse.  Thoughts, @dasilva333? Should I add it make a pull request with it added to the move-popup?

**Comment by dasilva333** [comment]
I don't mind pulling the request if you don't mind writing the code, to be clear you would go to View > General > click on a consumable (say Heavy Ammo Synth), click any character > move pop up will show up prompting the user to transfer 24 amount at that point it will look like this:
(lets assume this is the total amount)

Transfer Amount: 24 
-----------------------------
You can also choose to divide "Heavy Ammo Synth" evenly across all your profiles (8/8/8)

[ Transfer ] [ Normalize ] [ Cancel ]

**Comment by Mulchman** [comment]
Hadn't thought of it like that. I was thinking about adding it to the move popup, like this: http://imgur.com/94ikcN8

(mockup in paint, so excuse terrible-ness)

**Comment by Mulchman** [comment]
Or, what about the classes added to the new move popup icon: http://imgur.com/RQVlbI9

![Example](http://i.imgur.com/RQVlbI9.png)

**Comment by dasilva333** [comment]
Not really, Ive thought about adding more icons to the move pop up but due to the mobile limitation only fitting 6 icons I haven't added new ones like swap. However mats & consumables only ever take up 3 boxes max so a 4th wouldn't be so bad. I'll leave it up to you to decide whatever you think is better/easier to write.

Good luck bro

**Comment by Mulchman** [comment]
Created https://github.com/dasilva333/TowerGhostForDestiny/pull/140 with some of my thoughts on reworking things.

**Comment by Adiel-dev** [comment]
While this feature is very useful. It's complicated at frist to get a grip on how it works. Great that there's a dedicated "normalise" button(now called "extras" for some reason) but it still raises some questions if you never used it before:

What does it mean?
What items does it apply to?
What steps do I have to take in order to use the feature?

**Comment by Mulchman** [comment]
@Adiel-dev it is a bit technical sounding; I agree. In my fork I've cleaned up the text slightly, which might help? (Ignore the other option for 'normalize everything' as it's WIP.)

![example](http://i.imgur.com/nhuyvmg.png)

As for the 'extras' button in the move popup, that was done to not clutter up the move popup with tons and tons of buttons. Right now it's only used if you click on a consumable/materials, but @dasilva333 mentioned having plans to use it based on a different context when HoW drops.

**Comment by Adiel-dev** [comment]
"Is equally distribute xyz across your characters sufficient?"

Yes most definitely. But now you have to make sure people understand that "Normalize" and "Normalize Everthing" are actionable buttons. Now they kind of look like headers. And now comes the tricky part: how will you get people to use the feature? How will they know how to get to this screen? 

The "extra" button does suggest that there's more but only the most curious of people will ever push that button. 

Also the "Normalize Everything" actions seems more appropriate for the top menu instead of activating it through a separate and isolated/random material. When make it a top menu action I would ditch "Normalize" though.

**Comment by Mulchman** [comment]
Thanks, @Adiel-dev , good suggestions and comments.

**Comment by Mulchman** [comment]
Playing around a bit. Here's a mockup (some code + MS paint) of a distribute/split type dialog for consumables[1] (http://imgur.com/RI4dpif):

![example](http://i.imgur.com/RI4dpif.png)

Was thinking the workflow would be:
* hit a button or sub-button from the main TGfD view and this dialog pops up
* the initial view starts with all of your non-vault characters highlighted and their consumables[1]
* as you select or deselect characters, the item list would change depending on the consumables[1] those players had on them currently
* you then select which consumables[1] to distribute/split
* you then hit a button next to close (not pictured) to start the process

[1] picked consumables at random; dialog can/will include an option to select consumables or materials or both.

**Comment by dasilva333** [comment]
That's pretty good I like the way it is now but I say this is a great addition and makes using it a lot easier than having to pick each one out. What do you think @Adiel-dev @lanejcf @bluetidepro

**Comment by Mulchman** [comment]
There's something else I want to tie into this that I didn't mention. I think there needs to be 2 'modes': one for the distribution/split type of thing and one for easy and/or multiple character bulk transfer. Something I find myself doing a lot of, is individually transferring stacks of materials from one character to another because I need to level up a gun or piece of armor. I want an easy way to say 'give my warlock all my spinmetal'. I think this can be tied into this GUI, too. I'll make another mockup, probably tomorrow PDT, and post again.

**Comment by Mulchman** [comment]
New example (http://imgur.com/z9p1H6v):

![example](http://i.imgur.com/z9p1H6v.png)

(Apologies for the scaling; working from laptop and had to zoom out to fit the entire dialog on the screen.)

@dasilva333 @Adiel-dev @bluetidepro @lanejcf

**Comment by dasilva333** [comment]
It's good man I think this one gives you even more options but I think it's a bit hard to understand I showed it to my room mate and even after reading it was a bit unclear as to what to do. I think this might just be a verbiage issue but I'll let the real designers handle it. Thanks again @Mulchman

**Comment by bluetidepro** [comment]
I feel like this is way overly complicated. What about setting it up as more of a table view like this:

http://bluetide.pro/121Hp/5oyulMQK

The idea is you have a row for each material (and it shows the total of the item you have across the characters), and then each input would default `val` to the amount that character currently holds. Then a far right button icon that would equally distribute it across the characters. So the idea is you can quickly set the values for each character then there would be a proceed button at the button to actually submit it.

I think that is a WAY less complicated UX, and would be much easier to follow/do.

Thoughts @dasilva333 @Adiel-dev @bluetidepro @lanejcf?

**Comment by dasilva333** [comment]
@bluetidepro with up/down ticks I think it would make it a lot better. It's easy to follow and gives you the most control but I dont know if it fits the use case @Mulchman had in mind. I honestly am partial to any of these ideas I think they're all good ultimately it just has to work and be easy to understand right away.

Keep up the good work guys

**Comment by bluetidepro** [comment]
@dasilva333 Maybe I missed the use case @Mulchman had then, what was the specific use case you are referring to?

**Comment by dasilva333** [comment]
The quote is from this:

>  Something I find myself doing a lot of, is individually transferring stacks of materials from one character to another because I need to level up a gun or piece of armor. I want an easy way to say 'give my warlock all my spinmetal'

I think if we load up your form with that intent in mind the Spinmetal Row will probably have a default vaule of the balanced amount, so to get it to the max amount it would take either a change of the # manually or a lot of up ticks to get it to max, perhaps a double up arrow to represent the max value to set? I'm not sure I just think mobile wise setting the input field to max might be problematic.

**Comment by matthubb** [comment]
Regarding the **"I need upgrade mats for this item"** use case. 

I'm working on a specific feature for transferring the exact materials required for available upgrades on a given item.

**Comment by bluetidepro** [comment]
@dasilva333 I see what you mean now. How about next to the inputs, there is a button with an icon that would signify `max` or something. That way the user can quickly `max` out any of the inputs for a character. That would solve that problem! :)

**Comment by Mulchman** [comment]
Thanks for the comments thus far! I really do appreciate the feedback and ideas/criticisms. I'll be calling the pic above, http://imgur.com/z9p1H6v, "mega pic". It didn't take a lot of time to put that together, so don't feel like you can't say what you want because some time has already been invested or something. I did try to model it a bit after the loadouts UI, but it appears I failed, lol. ;)

> @dasilva333 Maybe I missed the use case @Mulchman had then, what was the specific use case you are referring to?

I'll try and do a big picture approach of how I use TGfD and/or what I want it to be able to do.

I have 3 different characters and usually only play a certain one each night. This leads to inventories getting lopsided/skewed with different types of items and what not. So, after a night of playing, I want to "normalize"/"equalize"/"evenly distribute" all of the player inventories with a single button press (instead of going through each item manually and doing the single-item-normalize process that currently exists in app). There are also times a player is short on a particular type of item so I want an easy button for the "give me all spinmetal" case.

Some of the ways to accomplish these things now with the app are really awkward to do on my Android tablet, which is what I use when out in the room that has the TV & PS4, due to screen space or slowness of the device or just the general awkwardness of interacting with a popup keyboard. As an example, a worse case attempt to normalize all consumables and materials on my players requires ~85 button presses plus ~17 app.refresh() waits (breakdown: clicking 17 unique consumables and materials + clicking "extras" + clicking "normalize" + clicking "ok" + waiting on transfers to take place + waiting on app.refresh() + clicking "ok"). The "give me all spinmetal" is a tad easier with "Auto Transfer Stacks" enabled but still requires multiple clicks on multiple players plus some app.refresh()es as stack sizes get out of sync and what not and it becomes impossible to transfer without doing an app.refresh()..

Recap of features I want TGfD to have:

1. equally distribute a single item across all players excluding the vault
  * This is in the app currently but @Adiel-dev brought up that it's kind of hidden and its intent, or what it will do, isn't well known/described/intimated
  * I don't want the vault included due to how I store my items in game but I figure others do, which is why it's in "mega pic"
2. equally distribute all "items" (consumables, materials, or both) across all players excluding the vault
  * Not in the app now, but what I'm trying to solve in the "mega pic"
  * The vault inclusion in the pic is to appease players that would want to include the vault in this feature
3. give me all of a single type of item (eg. give me all spinmetal), grabbing from all players including the vault
  * Not in the app now, but what I'm trying to solve in the "mega pic"

Some of my goals with coming up with the "mega pic" and the features I want to add are:

* eliminate any keyboard input
  * this helps mitigate any error checking code from user mistakes
  * it's awkward on mobile and significantly reduces screen real estate when the keyboard is out
* solve both of the things I wanted at once
  * this helps with menu entry bloat. A single entry wraps all the "normalize all consumables"/"normalize all materials"/"normalize all consumables and materials"/"bulk transfer items"/etc. type combinations
* move the visibility of the features to the top level app menu (not pictured/shown anywhere)

I think I rambled enough on that stuff.

~~--------------------------~~

> @matthubb said: I'm working on a specific feature for transferring the exact materials required for available upgrades on a given item.

That sounds awesome! I'd be interested to hear where you were going to place this functionality UI-wise. Should there be a top level menu bar option for "Utilities"/"Power Tools" or something where these non-standard behaviors can fit in (http://imgur.com/NhYduIP) or no?

![example](http://i.imgur.com/NhYduIP.png)

~~--------------------------~~

> @bluetidepro said: http://bluetide.pro/121Hp/5oyulMQK (just a rough idea)

> Those inputs could even be input spinners so you can tick up/down to increase/decrease the amounts. Also, the inputs would obviously have max & min attributes that didn't allow you to have more/less than the total amount in the 3 inputs.

Definitely a fan of not having a keyboard pop out on mobile.

**Comment by bluetidepro** [comment]
@Mulchman Okkkkayyy. Now I'm on the same page. So I def think we can use this http://bluetide.pro/121Hp/5oyulMQK as a good starting point to cover all the features you mentioned. Might need to add a few other buttons that describe what you mention in a one click/tap motion, but I think that's the best overall idea/layout. And then just added the ability to make it more visible in the nav. Right?

**Comment by M3Rocket** [comment]
While a table solution makes sense to cover complicated scenarios (normalize/move item/items from selected vault/characters to selected vault/characters), I think it can also make the interface unnecessarily cluttered and complicated. 

Items can be simplified by logical grouping. For instance, Spirit Bloom, Spinmetal, Relic Iron, and Helium can be grouped as upgrade materials. All the energies/shards can go in one group. And so on. Items that cannot be stacked like Crux of Crota can be left out. 

Having to type in numbers is slow on a phone or tablet in a table-like layout as depicted. Even clicking on tiny up-/down-tick arrows can be annoying.

Meanwhile though, I have an idea with regard to the current dialog box and how to make it more flexible without adding a bunch of complexity:

![tgd equalize](https://cloud.githubusercontent.com/assets/3769123/7597147/5ca25062-f8a6-11e4-9a32-2325cb847d7f.png)

The default idea is that if you wanted to equalize items between selected characters, the "from characters" would be the same as the "to characters." Therefore, the number of the item in the top row would change to reflect the total of that item in the checked vault/characters below.

**Comment by M3Rocket** [comment]
@matthubb @bluetidepro @Mulchman Any thoughts on my idea above for the normalize dialog box? I tried to stay consistent with regard to the current UI.

**Comment by Mulchman** [comment]
@M3Rocket I liked it. My first thought was that I don't know if the dialogs can have images in their titles? Join us on slack; there's a equalize_normalize channel on there.

**Comment by M3Rocket** [comment]
Latest iteration just in case anyone reading here rather than Slack has any other suggestions:

![tgd equalize](https://cloud.githubusercontent.com/assets/3769123/7639013/ddeb8908-fa2d-11e4-9122-fdcf14e038c9.png)

**Comment by dasilva333** [comment]
What does the Advanced.. button do?

**Comment by M3Rocket** [comment]
That "Advanced..." button is likely going to lead to a separate dialog with a much more complicated options for distributing, merging, consolidating, etc. materials between all characters in batch mode rather than one material at a time. That complicated setup could take awhile to flesh out, so "Advanced..." can easily be left out for now.

**Comment by Mulchman** [comment]
> @M3Rocket Latest iteration just in case anyone reading here rather than Slack has any other suggestions

Appreciate it. I hope to get to it soon, but real work has been busy.

**Comment by Mulchman** [comment]
Here's where I'm at with a redesign based on @M3Rocket 's comments-

Collapsed (http://imgur.com/Zlqo9bD):
![collapsed](http://i.imgur.com/Zlqo9bD.png)

Expanded (http://imgur.com/KJVEvin):
![expanded](http://i.imgur.com/KJVEvin.png)

The big matrix-of-options-thing (been talked about in the slack channel) will be in a more general case thing separate from this. This is only intended for normalizing a single consumable/material across user selected players.

**Comment by dasilva333** [comment]
@Mulchman I like it, it's very clean and well thought out good job so far

**Comment by M3Rocket** [comment]
@Mulchman Looks good to me--very nice! My preference is for it to auto-expand have it auto-expand for now until there are multiple "Extras Functions" sections.

**Comment by Mulchman** [comment]
@M3Rocket Sounds good. I also commented the button out until more "Extras Functions" exist (if ever).

**Comment by M3Rocket** [comment]
So nice and simple in 2.7.0! Thanx @Mulchman
